feat(httpserver-go1.22-redis): Inter-app communication (#311)
kraftcloud/examples: httpserver-go1.22-redis/README.md

Reviewer: ```suggestion
# Go and Redis HTTP Server
```

   You need a [BuildKit](https://github.com/moby/buildkit) builder. The easiest way to get one is via [Docker](https://docs.docker.com/engine/install/).
   Alternatively, you can also directly set up and use BuildKit, see the [quick start](https://github.com/moby/buildkit#quick-start).

> **Note**:
> The unikraft CLI is the current standard, while kraft is the legacy version.

> > Choose one of the CLIs below and only run the commands associated with it for the rest of this guide.

Reviewer: ```suggestion
   > **Note**:
   > The unikraft CLI is the current standard, while kraft is the legacy version.
   > Choose one of the CLIs below and only run the commands associated with it for the rest of this guide.
```

2. Clone the [`examples` repository](https://github.com/unikraft-cloud/examples) and `cd` into the `examples/httpserver-go1.22-redis` directory:

```bash
git clone https://github.com/unikraft-cloud/examples
cd examples/httpserver-go1.22-redis/
```

Make sure to log into Unikraft Cloud and pick a [metro](https://unikraft.com/docs/platform/metros) close to you.
This guide uses `fra` (Frankfurt, 🇩🇪):

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"
unikraft login
```

or

**Using the legacy kraft CLI**
```bash title="kraft"
# Set Unikraft Cloud access token
export UKC_TOKEN=token
# Set metro to Frankfurt, DE
export UKC_METRO=fra
```

> ### Deploy Redis

Reviewer: ```suggestion
## Deploy Redis
```

First, deploy the Redis instance.

> Redis is an internal service (not publicly accessible), reached via the `httpserver-go122-redis-redis.internal` domain.

Reviewer: that domain is brutal, maybe shorten it

uuid:            8abe24f6-5670-4e9b-8955-8ec10f3bad21
state:           starting
image:           <my-org>/httpserver-go-122-redis-db
resources:
  memory:        256MiB
  vcpus:         1
service:
  name:          late-sound-rhboe98o
  uuid:          01953c15-8a2e-4c1a-ac24-98a9417c5aa2
  domains:
  - fqdn:        httpserver-go122-redis-redis.internal
networks:
- uuid:          e76ae319-d210-4566-ad75-baed32fc3d2b
  private-ip:    10.0.0.85
  mac:           12:b0:0a:00:00:55
timestamps:
  created:       just now
scale-to-zero:
  enabled:       true
  policy:        idle
  stateful:      true
  cooldown-time: 1s
```

or

**Using the legacy kraft CLI**
```ansi title="kraft"
[●] Deployed successfully!
 │
 ├───────── name: redis-qbu65
 ├───────── uuid: a4708586-3a7c-4422-b3ad-915306293ac6
 ├──────── metro: https://api.fra.unikraft.cloud/v1
 ├──────── state: starting
 ├─────── domain: httpserver-go122-redis-redis.internal
 ├──────── image: oci://unikraft.io/<my-org>/redis@sha256:0b007ec2f56194da3a468aa93f4a2b8f267b1726ef3780355fbadb228fdf5c23
 ├─────── memory: 256 MiB
 ├────── service: small-resonance-jvj2drgu
 ├─ private fqdn: redis-qbu65.internal
 └─── private ip: 10.0.0.173
```

> ### Deploy Go HTTP Server

Reviewer: ```suggestion
## Deploy Go HTTP Server
```

 │
 ├───────── name: httpserver-go-mpzaa
 ├───────── uuid: 53fb29b9-6507-47ec-bc7b-a17502aef8be
 ├──────── metro: https://api.fra.unikraft.cloud/v1
 ├──────── state: starting
 ├─────── domain: https://misty-wildflower-vuvgji1u.fra.unikraft.app
 ├──────── image: oci://unikraft.io/<my-org>/httpserver-go@sha256:b923ac1ea60f83bdc61564aff95b11bac8a9e084f62532d0e882484d26e99da9
 ├─────── memory: 256 MiB
 ├────── service: misty-wildflower-vuvgji1u
 ├─ private fqdn: httpserver-go-mpzaa.internal
 └─── private ip: 10.0.0.85
```

In this case, the instance names are `httpserver-go-122-redis-db-2xc9u` / `redis-qbu65` and `httpserver-go-122-redis-app-bnnnc` / `httpserver-go-mpzaa`.
They're different for each run.

To set a value in Redis via the Go server, use the URL from the `fqdn` field:

```bash
curl -X POST -d "key=my-key" -d "value=my-value" https://<FQDN>
```

```
Success!
```

You can then retrieve the value:

```bash
curl https://<FQDN>/?key=my-key
```

```
Key "my-key" has value "my-value"
```

You can list information about the instances by running:

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"
unikraft instances list
```

```ansi title="unikraft"

> METRO  NAME                               STATE    IMAGE                             ARGS  MEMORY  VCPUS  FQDN                                         CREATED

Reviewer: pls align

fra    httpserver-go-122-redis-app-bnnnc  standby  <my-org>/httpserver-go-122-redis-app        256MiB  1      frosty-cherry-32qs6na2.fra.unikraft.app  4 minutes ago
fra    httpserver-go-122-redis-db-2xc9u   standby  <my-org>/httpserver-go-122-redis-db         256MiB  1      httpserver-go122-redis-redis.internal    5 minutes ago
```

or

**Using the legacy kraft CLI**
```bash title="kraft"
kraft cloud instance list
```

```ansi title="kraft"
NAME                 FQDN                                    STATE    STATUS        IMAGE                                   MEMORY   VCPUS  ARGS  BOOT TIME
httpserver-go-mpzaa  misty-wildflower-vuvgji1u.fra.unikr...  running  since 3secs   oci://unikraft.io/<my-org>/httpserver-g...  256 MiB  1            968.01 ms
redis-qbu65          httpserver-go122-redis-redis.internal   running  since 35secs  oci://unikraft.io/<my-org>/redis@sha256...  256 MiB  1            1707.20 ms
```

> ### Clean up

Reviewer: ```suggestion
## Cleanup
```

When done, remove the instances:

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"

> unikraft instances delete httpserver-go-122-redis-db-2xc9u httpserver-go-122-redis-app-bnnnc

Reviewer: why the names are so different than the ones below with `kraft`?

cd examples/httpserver-go1.22-redis/
```

Make sure to log into Unikraft Cloud and pick a [metro](https://unikraft.com/docs/platform/metros) close to you.
This guide uses `fra` (Frankfurt, 🇩🇪):

```bash title="unikraft"
unikraft login
```

or

```bash title="kraft"
# Set Unikraft Cloud access token
export UKC_TOKEN=token
# Set metro to Frankfurt, DE
export UKC_METRO=fra
```

> 3. Build and deploy the Redis instance:

Reviewer: Remove list numbering from here onwards

```bash title="unikraft"

> unikraft build redis --output <my-org>/redis:latest

Reviewer: ```suggestion
unikraft build ./redis --output <my-org>/redis:latest
```

unikraft run --scale-to-zero policy=idle,cooldown-time=1000,stateful=true --metro fra -m 256M --image <my-org>/redis:latest --domain httpserver-go122-redis-redis.internal
```

or

```bash title="kraft"

> kraft cloud deploy --scale-to-zero idle --scale-to-zero-stateful --scale-to-zero-cooldown 1s -M 256Mi --domain httpserver-go122-redis-redis.internal redis

Reviewer: ```suggestion
kraft cloud deploy --scale-to-zero idle --scale-to-zero-stateful --scale-to-zero-cooldown 1s -M 256Mi --domain httpserver-go122-redis-redis.internal ./redis
```

```

The output shows the Redis instance details:

```ansi title="kraft"
[●] Deployed successfully!
 │
 ├───────── name: redis-0zknj
 ├───────── uuid: 6363d50d-b598-4d25-84d2-bad936d5c9d6
 ├──────── metro: https://api.fra.unikraft.cloud/v1
 ├──────── state: starting
 ├─────── domain: httpserver-go122-redis-redis.internal
 ├──────── image: oci://unikraft.io/<my-org>/redis@sha256:390b45646a7840552c0f32aa3367de58e1774c7c93369847141ba61fa9d74ef2
 ├─────── memory: 256 MiB
 ├────── service: spring-haze-pn9mrw0l
 ├─ private fqdn: redis-0zknj.internal
 └─── private ip: 10.0.0.29
```

or

```ansi title="unikraft"
metro:         fra

> name:          redis-nqugb

Reviewer: ```suggestion
name:          redis-0zknj
```

> uuid:          3953b8b5-6bdd-4667-bf64-74a13f1450fa

Reviewer: ```suggestion
uuid:          6363d50d-b598-4d25-84d2-bad936d5c9d6
```

state:         starting
image:         <my-org>/redis
resources:
  memory:      256MiB
  vcpus:       1
service:
  uuid:        8b812ace-0449-46c2-bd32-a754cb871e04

>   name:        cool-smoke-xr4ykcoo

Reviewer: ```suggestion
  name:        spring-haze-pn9mrw0l
```

  domains:
  - fqdn:      httpserver-go122-redis-redis.internal
networks:
- uuid:        9930da73-d390-443c-85d9-200bef4e6fcf
  private-ip:  10.0.0.29
  mac:         12:b0:0a:00:00:1d
timestamps:
  created:     just now
scale-to-zero: policy=idle,stateful=true,cooldown-time=1s
```

4. Build and deploy the Go HTTP server:

```bash title="unikraft"

> unikraft build httpserver-go/ --output <my-org>/httpserver-go:latest

Reviewer: ```suggestion
unikraft build ./httpserver-go --output <my-org>/httpserver-go:latest
```

unikraft run --scale-to-zero policy=on,cooldown-time=1000 --metro fra -p 443:8080/tls+http -m 256M --image <my-org>/httpserver-go:latest --env REDIS_ADDR=httpserver-go122-redis-redis.internal:6379
```

or

```bash title="kraft"

> kraft cloud deploy --scale-to-zero on --scale-to-zero-cooldown 1s -p 443:8080/tls+http -M 256Mi httpserver-go/ --env REDIS_ADDR=httpserver-go122-redis-redis.internal:6379

Reviewer: ```suggestion
kraft cloud deploy --scale-to-zero on --scale-to-zero-cooldown 1s -p 443:8080/tls+http -M 256Mi ./httpserver-go --env REDIS_ADDR=httpserver-go122-redis-redis.internal:6379
```

```

The output shows the instance address and other details:

```ansi title="kraft"
[●] Deployed successfully!
 │
 ├───────── name: httpserver-go-mvw0b
 ├───────── uuid: a0e004cc-e6c1-4551-89b7-db8ce2e90ff3
 ├──────── metro: https://api.fra.unikraft.cloud/v1
 ├──────── state: starting
 ├─────── domain: https://bitter-brook-r4l6xl5h.fra.unikraft.app
 ├──────── image: oci://unikraft.io/<my-org>/httpserver-go@sha256:982204086be4ca18acad8e870a57577e206759cb0d572d1d1323d90aeb574307
 ├─────── memory: 256 MiB
 ├────── service: bitter-brook-r4l6xl5h
 ├─ private fqdn: httpserver-go-mvw0b.internal
 └─── private ip: 10.0.0.5
```

or

```ansi title="unikraft"
metro:         fra

> name:          httpserver-go-d0lzp

Reviewer: ```suggestion
name:          httpserver-go-mvw0b
```

> uuid:          8f44f094-4452-4d8a-b147-153e70492247

Reviewer: ```suggestion
uuid:          a0e004cc-e6c1-4551-89b7-db8ce2e90ff3
```

state:         starting
image:         <my-org>/httpserver-go
resources:
  memory:      256MiB
  vcpus:       1
service:
  uuid:        33aa78d6-c5ff-4e7a-bee0-5f486aa31f90

>   name:        misty-tree-j17ig1pp

Reviewer: ```suggestion
  name:        bitter-brook-r4l6xl5h
```

  domains:

>   - fqdn:      misty-tree-j17ig1pp.fra.unikraft.app

Reviewer: ```suggestion
  - fqdn:      bitter-brook-r4l6xl5h.fra.unikraft.app
```

networks:
- uuid:        bdf570af-7224-48f2-b374-b264b99c9d81
  private-ip:  10.0.0.5
  mac:         12:b0:0a:00:00:05
timestamps:
  created:     just now
scale-to-zero: policy=on,cooldown-time=1s
```

> In this case, the instance names are `redis-nqugb` and `httpserver-go-d0lzp`.

Reviewer: ```suggestion
In this case, the instance names are `redis-0zknj` and `httpserver-go-mvw0b`.
```

They're different for each run.

To set a value in Redis via the Go server, use the URL from the `domain` / `fqdn` field:

```bash

> curl -X POST -d "key=my-key" -d "value=my-value" https://<FQDN>

Reviewer: ```suggestion
curl -X POST -d "key=my-key" -d "value=my-value" https://bitter-brook-r4l6xl5h.fra.unikraft.app
```

Author: Do you mind if I keep the FQDN placeholder here? If someone is copying/pasting these commands quickly, the <FQDN> will fail more obviously

```

```
Success!
```

You can then retrieve the value:

```bash

> curl https://<FQDN>/?key=my-key

Reviewer: ```suggestion
curl https://bitter-brook-r4l6xl5h.fra.unikraft.app?key=my-key
```

```

```
Key "my-key" has value "my-value"
```

> 

Reviewer: Add listing output for the instances:
`You can list information about the instances by running: ... `

When done, you can remove the instances:

```bash title="unikraft"
unikraft instances delete redis-nqugb

> unikraft instances delete httpserver-go-d0lzp

Reviewer: ```suggestion
unikraft instances delete redis-0zknj httpserver-go-mvw0b
```

```

or

```bash title="kraft"
kraft cloud instance remove redis-nqugb

> kraft cloud instance remove httpserver-go-d0lzp

Reviewer: ```suggestion
kraft cloud instance remove redis-0zknj httpserver-go-mvw0b
```
